### ISSUE TITLE
Isolate pixel_wrapper from Pixel so we can use pixel as a submodule

### ISSUE THREAD

**Issue by EricHanLiu** [DDMAL/pixel_wrapper#10]
- Need to isolate all instances of pixel_wrapper code from the pixel.js source into a separate file 
  - use diff with the Pixel.js and pixel_wrapper repositories 
- Implement Pixel as a submodule and import everything properly

**Comment by EricHanLiu** [comment]
I'm pausing further work on this since it's not as important as getting the background layer working properly (another problem arose).

**Comment by EricHanLiu** [comment]
Tried exposing Pixel as a library and interacting with `module.export`ed functions, but the main issue remains instantiating the pixelInstance at the appropriate time and making sure it exists when trying to create the wrapper elements. I even check for if the instance is undefined, but there seems to be an issue with that since it passes as defined yet throws an undefined console error...

[Branch](https://github.com/DDMAL/pixel_wrapper/tree/separate_pixel)

**Comment by EricHanLiu** [comment]
Update!

I can just isolate all methods into a PixelWrapper class and instantiate this class in the `pixel.js` code when the plugin is activated (`activatePlugin()`), and then write some script that injects this exact code so we don't have to actually alter the standalone Pixel.

We'd just have the new `pixel_wrapper.js` file which contains all the Rodan-related methods, and a script which injects the instantiation code into `pixel.js` once it's imported as a submodule (and we can just add documentation to run this script when setting up pixel_wrapper). 

The file structure could look something like this
```
pixel_wrapper/source/js/plugins/
  -> Pixel.js/
    -> pixel_wrapper.js (as the only file)
```
Then we use the git submodule somehow within this `Pixel.js` directory and run the script.

**Comment by vigliensoni** [comment]
Excellent!

**Comment by EricHanLiu** [comment]
I'm gonna stop progress on the [separate_pixel_webpack](https://github.com/DDMAL/pixel_wrapper/tree/separate_pixel_webpack) branch which attempted to separate `pixel_wrapper` elements by interacting with the webpacked file (as a library) since it mostly didn't work.

Further work will be done on the [separate_pixel_source](https://github.com/DDMAL/pixel_wrapper/tree/separate_pixel_source) branch, where I'll try the method mentioned in my last comment.

**Comment by napulen** [comment]
Sounds good, any idea on how to perform the injection?

**Comment by EricHanLiu** [comment]
Maybe a python script that parses through the `pixel.js` file and inserts the few lines needed where necessary! 

As of now all `pixel_wrapper` methods are separated into a file `pixel-wrapper.js` inside the plugins/ folder. I'll start with cloning the standalone Pixel into my plugins/ folder and see if I can interact with that, then eventually turn that into a submodule. 

I'll have to edit one portion of Pixel though, but it won't affect the standalone version (it only affects things once it's imported in the wrapper).

**Comment by napulen** [comment]
I would encourage finding a solution within javascript. I am not familiar with this, but for the fragment of code you showed me, this looks like it could be solved with a "callback" function. If you can by some way get a variable that points to the pluginActivate() function, then you can make another function (sort of a wrapper, which we usually know as a callback) that calls the original one, and adds a few extra code afterwards.

Read this and let us know if it sounds reasonable.

http://javascriptissexy.com/understand-javascript-callback-functions-and-use-them/

**Comment by EricHanLiu** [comment]
Ok I've implemented Pixel.js as a git submodule in the plugins/ directory. I've manually added the few lines that needed to be changed for the wrapper to currently work, but I'll work on finding a way to inject them using JavaScript. 

I can push a separate branch to Pixel.js with the addition of ignoring the background layer for exportToData/CSV, and the extra code doesn't break the standalone Pixel.js either, and perhaps a pull request if the changes should be implemented in the standalone pixel? (this would make it easier to work with the wrapper)

OR what I can do is completely scrap the background button, and only have it generate on submitToRodan. This way I wouldn't need to change the standalone Pixel code, just the `pixel-wrapper.js` code, and it would only need to be generated once the user submits and Pixel.js is done.

**Comment by napulen** [comment]
So the button will not exist in standalone Pixel.js, am I following that correctly?

**Comment by napulen** [comment]
After the discussions I have seen lately, I am starting to think it would be productive to sit down and delimit what should and should **not** be in the standalone and wrapper versions. As of now, it seems blurry where the line is between the two in terms of features and implementation.

**Comment by EricHanLiu** [comment]
No extra buttons will exist in the standalone Pixel.js that don't already exist there. In the pixel_wrapper, I will only add the "Submit To Rodan" button, which will generate the background layer before submitting (pretty much how I had it originally).

I'm testing it now to see if the changes I've made are compatible with Rodan

**Comment by napulen** [comment]
Right. Sounds good.

**Comment by EricHanLiu** [comment]
Nestor and I tried today to see if we could omit having to use a script that inserts code into the `pixel.js` file, by interacting with the `pixel-wrapper.js` file and listening to an event that's thrown when Pixel is instantiated, and the eventListener then instantiates PixelWrapper.

However (and this seems kind of obvious in hindsight), without including the `pixel-wrapper.js` in `pixel.js`, `pixel-wrapper.js` never gets compiled by webpack (since `pixel.js` is the entrypoint). But we want to avoid having to include the `pixel-wrapper.js` file in the standalone Pixel, and changing the webpack entrypoint to `pixel-wrapper.js` also presents a plethora of other issues with Diva, so I think we'll have to resort to maybe a python script that adds the lines of code to the standalone Pixel file.

**Comment by EricHanLiu** [comment]
Set the default layers to 3, still not displaying the inputted layers correctly. They are uploaded, the proper RGB values are printed when I log the layer's colours, but they're not being displayed for some reason.

**Comment by EricHanLiu** [comment]
Fixed! It was an issue with the preBinarizedImageCanvas not being named the same thing in the standalone Pixel version.

**Comment by EricHanLiu** [comment]
I believe the default number of layers should be 3 to start, not just 1. It's set to 3 in the wrapper, and 1 in the standalone Pixel.js. 

@vigliensoni should I send a pull request to the Pixel repo to have this reverted back to having 3 layers to start, or should I just have the python script add the new layers so that there'll be 3 layers only in the wrapper Pixel.js? I think the standalone Pixel.js set the default to just 1 layer for generalizability, so maybe I shouldn't change that?

**Comment by EricHanLiu** [comment]
It's done!

Pixel.js is now a submodule with the latest commit version from master being linked to pixel_wrapper. All Rodan methods have been modified to work with the latest version of Pixel, and adapted to fit in the same class (as they were originally split across numerous classes). README instructions have been updated with the new/reordered commands for building everything properly.

The wrapper is activated with a python script which adds instantiation lines to the `pixel.js` file. This is the only method that I found worked (after trying the several methods listed above).

Exporting to Rodan can only be done if there are 3 layers present (an alert is sent otherwise), since we assume that now the CM requires 3 layers + the background. Otherwise, the job will crash if Rodan doesn't receive the right number of produced files (in this case 4 files)

After speaking with Jorge it seemed it'd be hard to generalize this number of input/output ports to change based on the number of Pixel layers, so that might be a future issue if we wish to generalize everything to **_n_ input/output ports** (though it should be doable).

[Pull request created.](https://github.com/DDMAL/pixel_wrapper/pull/23)
